db/migrate/20200427151946_drop_rackspace_cr.rb

```diff
@@ -0,0 +1,8 @@
+class DropRackspaceCr < ActiveRecord::Migration[5.2]
+  class Foreman::Model::Rackspace < ComputeResource
+  end
```

Comment:
This overrides:

```
  has_many_hosts
  has_many :hostgroups
  has_many :images, :dependent => :destroy
  has_many :compute_attributes, :dependent => :destroy
  has_many :compute_profiles, :through => :compute_attributes
 ```

I expect the migration to fail due to foreign keys if there is any content.

```
  add_foreign_key "compute_attributes", "compute_resources", name: "compute_attributes_compute_resource_id_fk"
  add_foreign_key "hostgroups", "compute_resources"
  add_foreign_key "hosts", "compute_resources", name: "hosts_compute_resource_id_fk"
  add_foreign_key "images", "compute_resources", name: "images_compute_resource_id_fk"
  add_foreign_key "key_pairs", "compute_resources", name: "key_pairs_compute_resource_id_fk"
```

I think we need to do a cascade delete for everything, except hosts and hostgroups which should be set to NULL.

Reply:
It should not, these are inherited from `ComputeResource`. However, we should probably first nullify hosts connected to it or the migration will fail due to `before_destroy EnsureNotUsedBy.new(:hosts)`

Reply:
Can you "nullify" both host and hostgroups? Images and Attributes do have "destroy" (cascade delete), however we don't have has_many statement for key_pairs. Any reason for that? We do not want to leave keys forever in the database.

Reply:
You actually brought interesting topic - I missed the fact you don't create a fake model, you actually use `ComputeResource` which can have some callbacks. Migration might fail. I think it should be safer to use `delete` instead of `destroy` on all associated models and the CR as well, what you think?

Reply:
I'm actually relying on the built-in callbacks to do the right thing here and nullify/destroy association, since there are so many associations. 
I've added nullifying hosts to the migration so it doesn't fail on the `before_destroy EnsureNotUsedBy.new(:hosts)` callback. 
There aren't key_pairs for rackspace, they is only used by GCE.

Reply:
Key pair is used also by Openstack (via `KeyPairComputeResource`) however you are right that Rackspace does not use it. I'd expect it to use them since OpenStack was founded by Raskspace, but I don't remember how this was implemented.

Reply:
Can you nullify hostgroups as well? The same problem, this can leave records referencing to non-existing ID in the database. There's not a foreign key for hostgroups for some reason but that does not mean we should allow this, it would probably fail at some point in Rails

Reply:
good point, i added `:dependent => :nullify` and explicitly nullified in the migration since it runs without a user, so the default hg taxonomy scope kicks in and hides the associated hostgroups.